mono/metadata/mono-route.c

```diff
@@ -0,0 +1,120 @@
+/*
+ * mono-route.c: Read the network routing tables using sysctl(3) calls
+ *               Required for Unix-like systems that don't have Linux's /proc/net/route
+ *
+ * Author:
+ *   Ben Woods (woodsb02@gmail.com)
+ */
+
+#if defined(PLATFORM_MACOSX) || defined(PLATFORM_BSD)
+
+#include <sys/socket.h>
+#include <net/if.h>
+#include <net/if_dl.h>
+#include <net/route.h>
+#include <netinet/in.h>
+#include <sys/param.h>
+#include <sys/sysctl.h>
+#include <stdlib.h>
+#include <string.h>
+#include <mono/metadata/object.h>
+#include <mono/metadata/mono-route.h>
+
+extern MonoBoolean ves_icall_System_Net_NetworkInformation_MacOsIPInterfaceProperties_ParseRouteInfo_internal(MonoString *iface, MonoArray **gw_addr_list)
+{
+	size_t needed;
+	in_addr_t in;
+	int mib[6];
+	int num_gws=0, gwnum=0;
+	unsigned int ifindex = 0;
+	char *buf, *next, *lim, *ifacename;
+	struct rt_msghdr *rtm;
+
+	MonoDomain *domain = mono_domain_get ();
+
+	ifacename = mono_string_to_utf8(iface);
+	if ((ifindex = if_nametoindex(ifacename)) == 0)
+		return FALSE;
+	g_free(ifacename);
+
+	// MIB array defining data to read from sysctl
+	mib[0] = CTL_NET;	// Networking
+	mib[1] = PF_ROUTE;	// Routing messages
+	mib[2] = 0;		// Protocol number (always zero)
+	mib[3] = AF_INET;	// Address family (IPv4)
+	mib[4] = NET_RT_DUMP;	// Dump routing table
+	mib[5] = 0;		//
+
+	// First sysctl call with oldp set to NULL to determine size of available data
+	if (sysctl(mib, sizeof(mib) / sizeof(mib[0]), NULL, &needed, NULL, 0) < 0)
+		return FALSE;
+
+	// Allocate suffcient memory for available data based on the previous sysctl call
+	if ((buf = malloc(needed)) == NULL)
+		return FALSE;
+
+	// Second sysctl call to retrieve data into appropriately sized buffer
+	if (sysctl(mib, sizeof(mib) / sizeof(mib[0]), buf, &needed, NULL, 0) < 0)
+		return FALSE;
+
+	lim = buf + needed;
+	for (next = buf; next < lim; next += rtm->rtm_msglen) {
+		rtm = (struct rt_msghdr *)next;
+		if (rtm->rtm_version != RTM_VERSION)
+			continue;
+		if (rtm->rtm_index != ifindex)
+			continue;
+		if((in = gateway_from_rtm(rtm)) == 0)
+			continue;
+		num_gws++;
+	}
+
+	*gw_addr_list = mono_array_new(domain, mono_get_string_class (), num_gws);
+
+	for (next = buf; next < lim; next += rtm->rtm_msglen) {
+		rtm = (struct rt_msghdr *)next;
+		if (rtm->rtm_version != RTM_VERSION)
+			continue;
+		if (rtm->rtm_index != ifindex)
+			continue;
+		if((in = gateway_from_rtm(rtm)) == 0)
+			continue;
+
+		MonoString *addr_string;
+		char addr [16], *ptr;
+
+		ptr = (char *) &in;
+		snprintf(addr, 16, "%u.%u.%u.%u",
```

Comment:
Please use sizeof(addr) in this call instead of the hard-coded 16. Should the size of addr ever change it will keep us from disaster in case if shrinking the array length. Also, do not ignore snprintf return value - check for truncated output and act accordingly if it happens